New official genie garage integration

swcloudgenie opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
This should entirely replace the existing `aladdin_connect` component, which is as of now broken.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is the official Genie garage integration, which will replace the existing broken integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
I have not made any attempt to deprecate or otherwise remove the existing `aladdin_connect` component. If you need assistance with this, just let me know the proper steps to take care of this. As of now, this PR simply adds the new working integration alongside the existing one.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
**NOTE**: I am not able to run the tests locally, they all fail because of bad imports. I am not a Python developer and am having trouble figuring out how to fix this issue.
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
**NOTE**: All testing was done manually with the external library

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

swcloudgenie committed:
new official genie garage integration

edenhaus commented on homeassistant/components/aladdin_genie/__init__.py:
```diff
@@ -0,0 +1,44 @@
+from . import api
+from .const import DOMAIN
+
+PLATFORMS: list[Platform] = [Platform.COVER]  # Platform.SENSOR (coming soon)
```
```suggestion
PLATFORMS: list[Platform] = [Platform.COVER]
```

edenhaus commented on homeassistant/components/aladdin_genie/__init__.py:
```diff
@@ -0,0 +1,44 @@
+    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+
+    # If using an aiohttp-based API lib
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = api.AsyncConfigEntryAuth(
```
Please use `entry.runtime_data` instead

edenhaus commented on homeassistant/components/aladdin_genie/__init__.py:
```diff
@@ -0,0 +1,44 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
```
```suggestion
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
```

edenhaus commented on homeassistant/components/aladdin_genie/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "dependencies": ["application_credentials"],
+  "documentation": "https://www.home-assistant.io/integrations/aladdin_genie",
+  "homekit": {},
```
```suggestion
```

edenhaus reviewed (changes requested):
Thanks for the contribution :)
Instead of creating a new integration, can you please modify the existing `aladdin_connect`?

swcloudgenie commented:
> Thanks for the contribution :) Instead of creating a new integration, can you please modify the existing `aladdin_connect`?

Our OAuth credentials have been stored under the domain `aladdin_genie`. In order to replace the existing integration, I will need these credentials to be stored under the `aladdin_connect` domain instead.

bemble commented:
> Our OAuth credentials have been stored under the domain `aladdin_genie`. In order to replace the existing integration, I will need these credentials to be stored under the `aladdin_connect` domain instead.

I created the credentials with `aladdin_connect` domain.

edenhaus commented on homeassistant/components/aladdin_connect/manifest.json:
```diff
@@ -1,11 +1,10 @@
   "domain": "aladdin_connect",
   "name": "Aladdin Connect",
-  "codeowners": ["@mkmer"],
+  "codeowners": ["@ajohnston1219"],
```
Why are you adding `@ajohnston1219` as code owner and not yourself?

swcloudgenie replied:
Nevermind, I think I managed to get this to work this time.

edenhaus commented on homeassistant/components/aladdin_connect/const.py:
```diff
@@ -1,22 +1,16 @@
-NOTIFICATION_ID: Final = "aladdin_notification"
-NOTIFICATION_TITLE: Final = "Aladdin Connect Cover Setup"
+DOMAIN = "aladdin_connect"
+VERSION = 2
```
```suggestion
CONFIG_FLOW_VERSION = 2
```

edenhaus commented on homeassistant/components/aladdin_connect/const.py:
```diff
@@ -1,22 +1,16 @@
-NOTIFICATION_TITLE: Final = "Aladdin Connect Cover Setup"
+DOMAIN = "aladdin_connect"
+VERSION = 2
+MINOR_VERSION = 1
```
```suggestion
CONFIG_FLOW_MINOR_VERSION = 1
```

edenhaus commented on homeassistant/components/aladdin_connect/sensor.py:
```diff
@@ -40,24 +41,6 @@ class AccSensorEntityDescription(SensorEntityDescription):
         state_class=SensorStateClass.MEASUREMENT,
         value_fn=AladdinConnectClient.get_battery_status,
     ),
-    AccSensorEntityDescription(
```
Why are these sensors being removed?

swcloudgenie replied:
Not all devices support this value, and there is no model number currently included in our partner API, therefore it requires modification of our API to support knowing whether or not to include this value.

edenhaus reviewed (dismissed):
The integration tests cannot bet setup. Can you please give it a look and fix it

swcloudgenie commented:
> The integration tests cannot bet setup. Can you please give it a look and fix it

I have fixed the import errors on my end. Now, when I run the tests, I get the following error:

```

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR at setup of test_stale_device_removal ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.fixture(scope="session", autouse=True)
    def prevent_ffmpeg_subprocess() -> Generator[None, None, None]:
        """Prevent ffmpeg from creating a subprocess."""
>       with patch(
            "homeassistant.components.ffmpeg.FFVersion.get_version", return_value="6.0"
        ):

tests/components/conftest.py:122:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/unittest/mock.py:1442: in __enter__
    self.target = self.getter()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'homeassistant.components.ffmpeg.FFVersion'

    def resolve_name(name):
        """
        Resolve a name to an object.

        It is expected that `name` will be a string in one of the following
        formats, where W is shorthand for a valid Python identifier and dot stands
        for a literal period in these pseudo-regexes:

        W(.W)*
        W(.W)*:(W(.W)*)?

        The first form is intended for backward compatibility only. It assumes that
        some part of the dotted name is a package, and the rest is an object
        somewhere within that package, possibly nested inside other objects.
        Because the place where the package stops and the object hierarchy starts
        can't be inferred by inspection, repeated attempts to import must be done
        with this form.

        In the second form, the caller makes the division point clear through the
        provision of a single colon: the dotted name to the left of the colon is a
        package to be imported, and the dotted name to the right is the object
        hierarchy within that package. Only one import is needed in this form. If
        it ends with the colon, then a module object is returned.

        The function will return an object (which might be a module), or raise one
        of the following exceptions:

        ValueError - if `name` isn't in a recognised format
        ImportError - if an import failed when it shouldn't have
        AttributeError - if a failure occurred when traversing the object hierarchy
                         within the imported package to get to the desired object.
        """
        global _NAME_PATTERN
        if _NAME_PATTERN is None:
            # Lazy import to speedup Python startup time
            import re
            dotted_words = r'(?!\d)(\w+)(\.(?!\d)(\w+))*'
            _NAME_PATTERN = re.compile(f'^(?P<pkg>{dotted_words})'
                                       f'(?P<cln>:(?P<obj>{dotted_words})?)?$',
                                       re.UNICODE)

        m = _NAME_PATTERN.match(name)
        if not m:
            raise ValueError(f'invalid format: {name!r}')
        gd = m.groupdict()
        if gd.get('cln'):
            # there is a colon - a one-step import is all that's needed
            mod = importlib.import_module(gd['pkg'])
            parts = gd.get('obj')
            parts = parts.split('.') if parts else []
        else:
            # no colon - have to iterate to find the package boundary
            parts = name.split('.')
            modname = parts.pop(0)
            # first part *must* be a module/package.
            mod = importlib.import_module(modname)
            while parts:
                p = parts[0]
                s = f'{modname}.{p}'
                try:
                    mod = importlib.import_module(s)
                    parts.pop(0)
                    modname = s
                except ImportError:
                    break
        # if we reach this point, mod is the module, already imported, and
        # parts is the list of parts in the object hierarchy to be traversed, or
        # an empty list if just the module is wanted.
        result = mod
        for p in parts:
>           result = getattr(result, p)
E           AttributeError: module 'homeassistant.components' has no attribute 'ffmpeg'

/usr/local/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pkgutil.py:528: AttributeError
```

This appears to be coming from a global test setup script. I am not sure how to resolve this. I have tried removing the venv and running the setup scripts again, but I am still getting this error. I have updated this branch with my most recent changes.

edenhaus commented:
Will check it tomorrow as I'm currently away from my notebook

swcloudgenie commented:
@edenhaus I have managed to fix the error above, I just had to install the test dependencies (I did not realize this was a separate step). However, I am now getting this error:

```
----------------------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------------------
INFO:homeassistant.loader:Loaded aladdin_connect from homeassistant.components.aladdin_connect
INFO:homeassistant.loader:Loaded application_credentials from homeassistant.components.application_credentials
INFO:homeassistant.loader:Loaded auth from homeassistant.components.auth
INFO:homeassistant.loader:Loaded websocket_api from homeassistant.components.websocket_api
INFO:homeassistant.loader:Loaded http from homeassistant.components.http
INFO:homeassistant.setup:Setting up http
INFO:homeassistant.setup:Setting up auth
INFO:homeassistant.setup:Setting up websocket_api
INFO:homeassistant.setup:Setting up application_credentials
INFO:homeassistant.setup:Setting up aladdin_connect
ERROR:homeassistant.config_entries:Error setting up entry Mock Title for aladdin_connect
Traceback (most recent call last):
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/config_entries.py", line 594, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/components/aladdin_connect/__init__.py", line 22, in async_setup_entry
    await config_entry_oauth2_flow.async_get_config_entry_implementation(
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/helpers/config_entry_oauth2_flow.py", line 428, in async_get_config_entry_implementation
    implementation = implementations.get(config_entry.data["auth_implementation"])
                                         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'auth_implementation'
------------------------------------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------------------------------------
2024-05-23 11:32:04.750 INFO     SyncWorker_0 homeassistant.loader:loader.py:764 Loaded aladdin_connect from homeassistant.components.aladdin_connect
2024-05-23 11:32:04.751 INFO     SyncWorker_0 homeassistant.loader:loader.py:764 Loaded application_credentials from homeassistant.components.application_credentials
2024-05-23 11:32:04.752 INFO     SyncWorker_0 homeassistant.loader:loader.py:764 Loaded auth from homeassistant.components.auth
2024-05-23 11:32:04.753 INFO     SyncWorker_0 homeassistant.loader:loader.py:764 Loaded websocket_api from homeassistant.components.websocket_api
2024-05-23 11:32:04.754 INFO     SyncWorker_0 homeassistant.loader:loader.py:764 Loaded http from homeassistant.components.http
2024-05-23 11:32:04.756 INFO     MainThread homeassistant.setup:setup.py:370 Setting up http
2024-05-23 11:32:04.759 INFO     MainThread homeassistant.setup:setup.py:370 Setting up auth
2024-05-23 11:32:04.761 INFO     MainThread homeassistant.setup:setup.py:370 Setting up websocket_api
2024-05-23 11:32:04.762 INFO     MainThread homeassistant.setup:setup.py:370 Setting up application_credentials
2024-05-23 11:32:04.763 INFO     MainThread homeassistant.setup:setup.py:370 Setting up aladdin_connect
2024-05-23 11:32:04.765 ERROR    MainThread homeassistant.config_entries:config_entries.py:678 Error setting up entry Mock Title for aladdin_connect
Traceback (most recent call last):
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/config_entries.py", line 594, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/components/aladdin_connect/__init__.py", line 22, in async_setup_entry
    await config_entry_oauth2_flow.async_get_config_entry_implementation(
  File "/Users/adam/projects/homeassistant/hass-core/homeassistant/helpers/config_entry_oauth2_flow.py", line 428, in async_get_config_entry_implementation
    implementation = implementations.get(config_entry.data["auth_implementation"])
                                         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'auth_implementation'

 tests/components/aladdin_connect/test_init.py ⨯                                                                                                                                                                             100% ██████████
========================================================================================================= short test summary info ==========================================================================================================
```

It looks like this is failing to set up the OAuth flow. Is there documentation anywhere on how to mock this OAuth flow for the tests?

balloob commented on tests/components/aladdin_connect/test_init.py:
```diff
@@ -41,102 +30,25 @@ async def test_setup_get_doors_errors(hass: HomeAssistant) -> None:
     hass: HomeAssistant, mock_aladdinconnect_api: MagicMock
 ) -> None:
     """Test loading and unloading Aladdin Connect entry."""
     config_entry = MockConfigEntry(
```
you are creating a config entry here that is not matching the data format that OAuth2 config entries follow. You can see the format in your test instance by checking the file `<config>/.storage/core.config_entries`

balloob commented on tests/components/aladdin_connect/test_config_flow.py:
```diff
@@ -1,278 +0,0 @@
-"""Test the Aladdin Connect config flow."""
```
All config flows need to be tested. OAuth2 config flow tests are automatically generated when you use the scaffold script. You can see the template here: https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow_oauth2/tests/test_config_flow.py

swcloudgenie replied:
Okay, I was trying to fix tests from the previous integration, but those appear to be incompatible. I have used the scaffold script to create a new integration and generate the tests. The tests are now passing.

balloob commented:
You have not removed the old test file `test_init.py`, which still references the no longer installed PyPI package `AIOAladdinConnect`. This is causing CI to fail. I would have fixed this for you, but you have disabled our option to make changes to your branch.

You can reproduce this locally by removing that package `python3 -m pip uninstall AIOAladdinConnect` and then run the tests.

For next updates, can you please stop force pushing and squash and merging your updates. It makes this PR harder to review as we can only do full reviews and not incremental ones.

swcloudgenie commented:
> You have not removed the old test file `test_init.py`, which still references the no longer installed PyPI package `AIOAladdinConnect`. This is causing CI to fail. I would have fixed this for you, but you have disabled our option to make changes to your branch.
> 
> You can reproduce this locally by removing that package `python3 -m pip uninstall AIOAladdinConnect` and then run the tests.
> 
> For next updates, can you please stop force pushing and squash and merging your updates. It makes this PR harder to review as we can only do full reviews and not incremental ones.

I believe I have fixed the tests. And yes, I will stop squashing the commit moving forward. I was trying to keep this change in a single commit, but I see your point about making the review process more difficult. Sorry for the trouble.

balloob commented on homeassistant/components/aladdin_connect/__init__.py:
```diff
@@ -1,48 +1,56 @@
-
-PLATFORMS: list[Platform] = [Platform.COVER, Platform.SENSOR]
+PLATFORMS: list[Platform] = [Platform.COVER]
+SCAN_INTERVAL = timedelta(seconds=15)
```
This is not used.

balloob commented on homeassistant/components/aladdin_connect/const.py:
```diff
@@ -1,22 +1,18 @@
 SUPPORTED_FEATURES: Final = CoverEntityFeature.OPEN | CoverEntityFeature.CLOSE
-CLIENT_ID = "1000"
+
+SCAN_INTERVAL = 20
```
Not used

swcloudgenie replied:
How do I set the scan interval for polling? The documentation says this:

> You can control the polling interval for your integration by defining a SCAN_INTERVAL constant in your platform. Careful with setting this too low. It will take up resources in Home Assistant, can overwhelm the device hosting the API or can get you blocked from cloud APIs. The minimum allowed value is 5 seconds.

```
from datetime import timedelta

SCAN_INTERVAL = timedelta(seconds=5)
```
Where should I define this constant?

balloob replied:
you need to define it inside `cover.py`

balloob replied:
The default scan interval for covers is every 15 seconds, which is what is being used now.

balloob commented on homeassistant/components/aladdin_connect/const.py:
```diff
@@ -1,22 +1,18 @@
-DOMAIN = "aladdin_connect"
+API_URL = "https://twdvzuefzh.execute-api.us-east-2.amazonaws.com/v1"
+API_KEY = "k6QaiQmcTm2zfaNns5L1Z8duBtJmhDOW8JawlCC3"
 SUPPORTED_FEATURES: Final = CoverEntityFeature.OPEN | CoverEntityFeature.CLOSE
```
Can be moved into `cover.py` and assigend directly, no need to push into const.

balloob commented:
@swcloudgenie 3 tiny comments. Ok to merge once addressed. I will make sure it is included in the next beta and will be published on June 5.

swcloudgenie committed:
move api constants into api module

swcloudgenie committed:
move scan interval constant to cover.py

balloob commented:
🐬  thank you 💯

joostlek commented on homeassistant/components/aladdin_connect/__init__.py:
```diff
@@ -1,48 +1,53 @@
+    """Migrate old config."""
+    if config_entry.version < CONFIG_FLOW_VERSION:
+        config_entry.async_start_reauth(hass)
+        new_data = {**config_entry.data}
+        hass.config_entries.async_update_entry(
+            config_entry,
+            data=new_data,
```
You can omit the data here, it doesnt add value to store it as a new dict

joostlek commented on homeassistant/components/aladdin_connect/api.py:
```diff
@@ -0,0 +1,31 @@
+        if not self._oauth_session.valid_token:
+            await self._oauth_session.async_ensure_token_valid()
+
+        return str(self._oauth_session.token["access_token"])
```
`cast(str, self._oauth_session.token["access_token"])` is better to use here as it doesn't create a new string

joostlek commented on homeassistant/components/aladdin_connect/config_flow.py:
```diff
@@ -1,137 +1,58 @@
-    except Aladdin.InvalidPasswordError as ex:
-        raise InvalidAuth from ex
 
+class OAuth2FlowHandler(
```
One major thing I am missing is, can we set some unique user id as unique_id for the config flow? This now creates the situation where someone can break their installation by logging in with the wrong account. We should prevent that.

joostlek commented on homeassistant/components/aladdin_connect/const.py:
```diff
@@ -1,22 +1,14 @@
+OAUTH2_TOKEN = "https://twdvzuefzh.execute-api.us-east-2.amazonaws.com/v1/oauth2/token"
 
-DOMAIN = "aladdin_connect"
 SUPPORTED_FEATURES: Final = CoverEntityFeature.OPEN | CoverEntityFeature.CLOSE
```
Can we move this to `cover.py`?

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
 ) -> None:
     """Set up the Aladdin Connect platform."""
-    acc: AladdinConnectClient = hass.data[DOMAIN][config_entry.entry_id]
+    session: api.AsyncConfigEntryAuth = config_entry.runtime_data
+    acc = AladdinConnectClient(session)
```
session is always used to create a `AladdinConnectClient`. Let's move this to `__init__.py` so we don't have to create this twice.

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
 ) -> None:
     """Set up the Aladdin Connect platform."""
-    acc: AladdinConnectClient = hass.data[DOMAIN][config_entry.entry_id]
+    session: api.AsyncConfigEntryAuth = config_entry.runtime_data
```
runtime_data should be used if the ConfigEntry is typed. Please type it,.

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
-    acc: AladdinConnectClient = hass.data[DOMAIN][config_entry.entry_id]
+    session: api.AsyncConfigEntryAuth = config_entry.runtime_data
+    acc = AladdinConnectClient(session)
     doors = await acc.get_doors()
```
We can move the door fetching to `__init__.py` to prevent it fetching the doors twice.

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
+    session: api.AsyncConfigEntryAuth = config_entry.runtime_data
+    acc = AladdinConnectClient(session)
     doors = await acc.get_doors()
     if doors is None:
         raise PlatformNotReady("Error from Aladdin Connect getting doors")
```
This can be removed

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
     doors = await acc.get_doors()
     if doors is None:
         raise PlatformNotReady("Error from Aladdin Connect getting doors")
+    device_registry = dr.async_get(hass)
+    doors_to_add = []
+    for door in doors:
+        existing = device_registry.async_get(door.unique_id)
+        if existing is None:
+            doors_to_add.append(door)
```
I am not sure what the idea is here, but currently you only add garage doors that have been added previously. This should be revistied and fixed

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -28,25 +26,33 @@ async def async_setup_entry(
+
     async_add_entities(
-        (AladdinDevice(acc, door, config_entry) for door in doors),
+        (AladdinDevice(acc, door, config_entry) for door in doors_to_add),
```
please remove the trailing comma

joostlek commented on homeassistant/components/aladdin_connect/cover.py:
```diff
@@ -74,74 +80,52 @@ class AladdinDevice(CoverEntity):
         if value is None:
             return None
-        return value == STATE_CLOSED
+        return bool(value == "closed")
```
This already is a bool

joostlek commented on homeassistant/components/aladdin_connect/model.py:
```diff
@@ -5,12 +5,26 @@
+    battery_level: int
+
+
+class GarageDoor:
```
I am not sure what the added benefit of this class is. It just wraps the class above and only sets an extra unique_id.

mkmer commented on homeassistant/components/aladdin_connect/__init__.py:
```diff
@@ -1,48 +1,53 @@
+from . import api
+from .const import CONFIG_FLOW_MINOR_VERSION, CONFIG_FLOW_VERSION
 
-PLATFORMS: list[Platform] = [Platform.COVER, Platform.SENSOR]
```
This will "strand" the sensors for the existing users.... should the update remove them?

joostlek replied:
Oh this is a major issue, this should be resolved as well

joostlek commented on tests/components/aladdin_connect/test_config_flow.py:
```diff
@@ -1,278 +1,82 @@
-        await hass.async_block_till_done()
+        "homeassistant.components.aladdin_connect.async_setup_entry", return_value=True
+    ) as mock_setup:
+        await hass.config_entries.flow.async_configure(result["flow_id"])
```
We should check the result of this call to check if the config entry is created with the right data/

joostlek reviewed (commented):
Please check the comments and fix them in follow up PRs

mkmer commented:
No docs update?  The credential screen instructs the user to follow the documentation for OAuth clientID and secret, but there is no new instructions or PR for them.

mkmer commented:
Should we remove the "app" notification and repair issue for the "old" version?

mkmer commented on homeassistant/components/aladdin_connect/__init__.py:
```diff
@@ -1,48 +1,53 @@
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
```
Should we be doing all the intermediate assignments for single use?
Remove `implementation` and `session` and inline the code?

joostlek replied:
Because they are long I don't mind this tbh, makes it a bit cleaner

mkmer replied:
I too like the easier read, but I think it's a startup performance hit that bdraco would go after :) (and has on several of my integrations)

joostlek replied:
Assigning to a variable in a temporary context isn't that bad tbf, I use it in Withings as well. Curious to those PRs of bdraco though!

balloob replied:
it's based on the scaffold template.
